src/html/validate-element-name.ts

```diff
@@ -0,0 +1,102 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import babelTraverse from 'babel-traverse';
+import * as babel from 'babel-types';
+import {Document, Severity, SourceRange, Warning} from 'polymer-analyzer';
+const validate = require('validate-element-name');
+
+import {registry} from '../registry';
+import {getDocumentContaining, stripIndentation} from '../util';
+
+import {Rule} from '../rule';
+
+class ValidateElementName extends Rule {
+  code = 'validate-element-name';
+  description = stripIndentation(`
+      Warns for invalid element names.
+  `);
+
+  async check(document: Document) {
+    const warnings: Warning[] = [];
+
+    const elements = document.getFeatures({kind: 'polymer-element'});
+    if (elements.size === 0) {
+      return warnings;
+    }
+
+    for (const el of elements) {
+      if (el.tagName === undefined) {
+        continue;
+      }
+
+      const containingDoc = getDocumentContaining(el.sourceRange, document);
+      if (containingDoc === undefined) {
+        continue;
+      }
+
+      const validationResult = validate(el.tagName);
+      if (validationResult.isValid && validationResult.message === undefined) {
+        continue;  // Valid element
+      }
+
+      const isP2 = babel.isClassDeclaration(el.astNode);
+      let sourceRange: SourceRange|undefined;
```

Comment:
let's have the entire element declaration as a fallback source range, in case we don't get a source range for the tagname itself (e.g. in the case of a totally vanilla custom element, or one based on a non-polymer library).

(btw, I really love the work you've done here to get a super specific source range, and that you only do it in the rare cases where there is a problem)

Reply:
Thanks 😄